### DISCUSSION TITLE
Generic ODBC SqlConnector

### DISCUSSION THREAD

**Discussion by ev1ls33d** [seaweedfs/seaweedfs#3493] [Ideas]
I'm not sure if this is easy to implement, since I'm not familiar with GO and it's libs, but would it be possible to provide a simple ODBC connector for filer?

I digged a little into the code and found the generic sql interface, which brought me to the conclusion, that the following properties would be enough to enable a connection:

- ConnectionString <connection string>
- SqlInsert <template string>
- SqlUpdate <template string>
- SqlDelete <template string>
- SqlCreateTable <template string>
- SqlDropTable <template string>
- SqlCommit / SqlRollback <template string>

Maybe some more templates for the other interface methods are required? I can't tell, but this should basically be enough to keep the metadata updated, right?

There is a ton of SQL databases that would be supported, if we had ODBC (which can be used on Windows and Linux afaik).

What is your opinion on this?

**Comment by chrislusf** [comment]
In general, a pure Go driver is desired. ODBC seems always need cGo. Maybe you want to use SQLServer? There are some pure Go drivers for it.

Hopefully, just something similar to https://github.com/seaweedfs/seaweedfs/blob/3afda0c89c8faa584423b6b1b488cd18d1d42a88/weed/filer/postgres2/postgres2_store.go is what you need to add a new SQL dialect.

**Reply by ev1ls33d** [reply]
Unfortunately I need to use it with a Sybase Anywhere Database.
It's similar to MSSQL (syntax wise) but not completely compatible.

I found this: https://github.com/alexbrainman/odbc
But as you mentioned it relies on either odbc32.dll or cGo. 
There also seems to be a Sybase Anywhere Go driver, but it relies on it's C API ( https://github.com/a-palchikov/sqlago ).

I guess, since both are not native go packages, odbc offers a broader spectrum of additional database backends.

**Reply by chrislusf** [reply]
https://golangexample.com/sap-formerly-sybase-ase-rs-iq-driver-written-in-pure-go/

**Reply by ev1ls33d** [reply]
ASE/RS/IQ != Anywhere unfortunately

**Reply by kmlebedev** [reply]
> ASE/RS/IQ != Anywhere unfortunately

And is it really justified?  Since there are vague doubts about the speed of such a connection and asynchronous requests?

Maybe a driver will suffice?  https://github.com/microsoft/go-mssqldb

**Reply by ev1ls33d** [reply]
I guess to answer those questions I'll have to read up about the TDS protocol which Sybase Anywhere also seems to support.
But my knowledge is not sufficient to be able to predict performance diffrences between an odbc connection and a compatible TDS implementation yet.

An approach to figure out the performance impact, would be a test setup with odbc (even if not pure) and a postgres native connection and measure the diffrence. 
But as stated above, I'm not familiar with go and I'm not sure if I have the time to dive into the language to test this myself.

**Reply by kmlebedev** [reply]
> I guess to answer those questions I'll have to read up about the TDS protocol which Sybase Anywhere also seems to support. But my knowledge is not sufficient to be able to predict performance diffrences between an odbc connection and a compatible TDS implementation yet.
> 
> An approach to figure out the performance impact, would be a test setup with odbc (even if not pure) and a postgres native connection and measure the diffrence. But as stated above, I'm not familiar with go and I'm not sure if I have the time to dive into the language to test this myself.


A new driver can be conditionally added in an evening, but this requires practical application, who will support such an installation in production?

**Reply by kmlebedev** [reply]
If you look towards the azure cloud, then you need an ms sql driver https://devblogs.microsoft.com/azure-sql/go-azure-sql/

**Reply by ev1ls33d** [reply]
I have a practical application that fails to scale and I'm thinking about using seaweed as an infrastructure to upgrade its DMS service implementation. 
Since there is no "file level" security in seaweed and the application has already a meta-data structure with access rights for it's current documents, I thought I could add some glue code to marry the two concepts and get a DMS service, which only shows files which the requesting (database-) user has access to.

**Reply by kmlebedev** [reply]
There is also an article where the odbc driver is recommended above.  https://www.progress.com/tutorials/odbc/connect-to-oracle-salesforce-etc.-using-odbc-from-go-on-windows
It seems you can try to run the benchmark conditionally for mysql or postgreSql via odbc connection.  But again, for this there must be practical applications and a real task.

**Reply by ev1ls33d** [reply]
> There is also an article where the odbc driver is recommended above. https://www.progress.com/tutorials/odbc/connect-to-oracle-salesforce-etc.-using-odbc-from-go-on-windows It seems you can try to run the benchmark conditionally for mysql or postgreSql via odbc connection. But again, for this there must be practical applications and a real task.

As described above, for me this task is quite real. If you're skeptical about the concept I'd gladly take advice to adjust my ideas.
At the end of the day I'd like a distributed FS to support the DMS service (certified!) which must guarantee to respect the users privileges.
 
To me it seemed the best to interact with the filer meta data in some way and filter it based on a jwd token or s.th. alike.
